specs/beacon-chain.md

```diff
@@ -791,13 +795,32 @@ def change_validators(validators: List[ValidatorRecord]) -> None:
             # STUB: withdraw to shard chain
 ```
 
-Finally:
+Then:
 
 * Set `crystallized_state.validator_set_change_slot = crystallized_state.last_state_recalculation_slot`
 * For all `c` in `crystallized_state.crosslinks`, set `c.recently_changed = False`
 * Let `next_start_shard = (shard_and_committee_for_slots[-1][-1].shard + 1) % SHARD_COUNT`
 * Set `shard_and_committee_for_slots[CYCLE_LENGTH:] = get_new_shuffling(active_state.randao_mix, validators, next_start_shard)`
 
+#### Finally...
+
+* For any validator with index `v` with balance less than `MIN_ONLINE_DEPOSIT_SIZE` and status `ACTIVE`, run `exit_validator(v, crystallized_state, penalize=False, current_slot=block.slot)`
+* Set `crystallized_state.last_state_recalculation_slot += CYCLE_LENGTH`
+* Remove all attestation records older than slot `crystallized_state.last_state_recalculation_slot`
+* Empty the `active_state.pending_specials` list
+* Set `active_state.recent_block_hashes = active_state.recent_block_hashes[CYCLE_LENGTH:]`
+* Set `shard_and_committee_for_slots[:CYCLE_LENGTH] = shard_and_committee_for_slots[CYCLE_LENGTH:]`
+
+For any validator that was added or removed from the active validator list during this state recalculation:
+
+* If the validator was removed, remove their index from the `current_persistent_committees` and `next_persistent_committees` objects.
+* If the validator was added with index `validator_index`, let `assigned_shard = hash(crystallized_state.randao_mix + bytes8(validator_index)) % SHARD_COUNT`. Add `validator_index` to the end of `next_persistent_committees[assigned_shard]`.
+
+If `block.slot % SHARD_PERSISTENT_COMMITTEE_CHANGE_PERIOD == 0`, then:
+
+* Set `current_persistent_committees = next_persistent_committees`
+* Set `next_persistent_committees = get_persistent_shuffling(validators, crystallized_state.randao_mix)`
```

Comment:
I thought we were going to slowly, but consistently, be shuffling small amounts of validators at each cycle (or validator change). So we never had _all_ of the validators changing at once. I see instead of going that route (which would certainly be more complicated algorithmically) you are instead changing all of the validators at once, but giving them a 12 day look ahead.

I suppose this is reasonable :)

Reply:
Could try to come up with a version that changes more gradually. Perhaps something like saving a list of "future shard change records", and every cycle adding a record for 12 days from now that moves a randomly selected active validator index to a random shard?

I'd totally be ok with that approach.